src/certifier.cc

```diff
@@ -25,444 +25,218 @@ using namespace certifier::utilities;
-    tagged_claim* t = psm.add_claims();
-    t->set_tag(tc_[i]->tag());
-    t->mutable_claim()->CopyFrom(tc_[i]->claim());
+  store_entry* certifier::framework::policy_store::get_entry(unsigned ent) {
+    if (ent >= num_ents_)
+      return nullptr;
+    return entry_[ent];
   }
-  for (int i = 0; i < num_tkm_; i++) {
-    channel_key_message* t = psm.add_channel_authentication_keys();
-    t->set_tag(tkm_[i]->tag());
-    t->mutable_auth_key()->CopyFrom(tkm_[i]->auth_key());
-  }
-  for (int i = 0; i < num_blobs_; i++) {
-    tagged_blob_message* t = psm.add_blobs();
-    t->CopyFrom(*tagged_blob_[i]);
-  }
-
-  if (!psm.SerializeToString(out))
-    return false;
-
-  return true;
-}
-
-bool certifier::framework::policy_store::Deserialize(string& in) {
-  policy_store_message psm;
 
-  if (!psm.ParseFromString(in))
-    return false;
-
-  // Copy data to psm
-  if (psm.has_policy_key()) {
-    policy_key_valid_ = true;
-    policy_key_.CopyFrom(psm.policy_key());
-  } else
-    policy_key_valid_ = false;
+  bool certifier::framework::policy_store::update_or_insert(const string& tag,
+              const string& type, const string& value) {
+    int ent = find_entry_by_tag_and_type(tag, type);
+    if (ent < 0) {
+      if (!add_entry(tag, type, value)) {
+        return false;
+      } else {
+        return true;
+      }
+    }
+    put(ent, value);
+    return true;
+  }
 
-  num_ts_ = psm.trusted_services_size();
-  for (int i = 0; i < num_ts_; i++) {
-    trusted_service_message* t = new(trusted_service_message);
-    t->set_tag(psm.trusted_services(i).tag());
-    t->CopyFrom(psm.trusted_services(i));
-    ts_[i] = t;
-  }
-  num_tsc_ = psm.signed_claims_size();
-  for (int i = 0; i < num_tsc_; i++) {
-    tagged_signed_claim* t = new(tagged_signed_claim);
-    t->set_tag(psm.signed_claims(i).tag());
-    t->mutable_sc()->CopyFrom(psm.signed_claims(i).sc());
-    tsc_[i]= t;
-  }
-  num_si_ = psm.storage_info_size();
-  for (int i = 0; i < num_si_; i++) {
-    storage_info_message* t = new(storage_info_message);
-    t->CopyFrom(psm.storage_info(i));
-    si_[i]= t;
-  }
-  num_tc_ = psm.claims_size();
-  for (int i = 0; i < num_tc_; i++) {
-    tagged_claim* t = new(tagged_claim);
-    t->set_tag(psm.claims(i).tag());
-    t->mutable_claim()->CopyFrom(psm.claims(i).claim());
-    tc_[i] = t;
-  }
-  num_tkm_ = psm.channel_authentication_keys_size();
-  for (int i = 0; i < num_tkm_; i++) {
-    channel_key_message* t = new(channel_key_message);
-    t->set_tag(psm.channel_authentication_keys(i).tag());
-    t->mutable_auth_key()->CopyFrom(psm.channel_authentication_keys(i).auth_key());
-    tkm_[i] = t;
-  }
-  num_blobs_ = psm.blobs_size();
-  for (int i = 0; i < num_blobs_; i++) {
-    tagged_blob_message* t = new(tagged_blob_message);
-    t->CopyFrom(psm.blobs(i));
-    tagged_blob_[i] = t;
+  void certifier::framework::policy_store::delete_entry(unsigned ent) {
+    if (num_ents_ <= 0)
+      return;
+    delete entry_[ent];
```

Comment:
Note, if checks suggested on prev line are not right, this will end up deleting some random out-of-bounds entry.

The L165 is also fraught with bad behaviour if say num_ents_ is 5 and `ent` is 10.

And, even worse, the code below will try shift entries up one cell but may never end the loop.